### PR TITLE
ISPN-14725 Hot Rod Client GraalVM Feature

### PR DESCRIPTION
**Pull request by ryanemerson**
https://issues.redhat.com/browse/ISPN-14725

### CODE REVIEW

File: graalvm/core/src/test/java/org/infinispan/client/hotrod/graalvm/core/EmbeddedTest.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package org.infinispan.integrationtests.graalvm.core;
+package org.infinispan.client.hotrod.graalvm.core;
```
Comment (pruivo) [inline graalvm/core/src/test/java/org/infinispan/client/hotrod/graalvm/core/EmbeddedTest.java:1]: is this change intentional?

Comment (ryanemerson) [reply graalvm/core/src/test/java/org/infinispan/client/hotrod/graalvm/core/EmbeddedTest.java]: No, good spot. It seems I also change the CDI integration tests as well. I have reverted these changes, but I have updated the EmbeddedTest package to match those found in `src/main`.

File: graalvm/client-hotrod/src/test/resources/reflect-config.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+[
+  {
+    "name": "org.testcontainers.shaded.com.github.dockerjava.core.DockerConfigFile",
```
[the remark this answers is not in the document]
Comment (ryanemerson) [reply graalvm/client-hotrod/src/test/resources/reflect-config.json:3]: > We are building test containers as native as well?

Maybe this is what you meant, but it's just the code that launches the containers. 

> We are building test containers as native as well? Shouldn't we just run tests normally and then load a native executable like quarkus does?

In Quarkus aren't the integration tests effectively just calling the APIs of the app consuming the client? Doing it this way we don't have to expose the client functionality, we can just test it direct in a native binary.

File: graalvm/client-hotrod/src/test/java/org/infinispan/client/hotrod/graalvm/substitutions/SubstituteClusteredIT.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+public class SubstituteClusteredIT {
+}
+
+@TargetClass(ClusteredIT.class)
+final class Target_ClusteredIT {
+
+   @Substitute
+   public static JavaArchive[] artifacts() {
+      return new JavaArchive[0];
```
[the remark this answers is not in the document]
Comment (ryanemerson) [reply graalvm/client-hotrod/src/test/java/org/infinispan/client/hotrod/graalvm/substitutions/SubstituteClusteredIT.java:17]: > Can we add some documentation here as to why this works? Where are all these classes and jars added?

They aren't :slightly_smiling_face:. With this approach we have to avoid tests that rely on server-side dependencies. It might be possible to get Shrinkwrap to work with native as expected, but I didn't have much time to investigate.

I've added a comment to the class.

File: graalvm/client-hotrod/pom.xml [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+                        <goals>
+                           <goal>test</goal>
+                        </goals>
+                        <phase>test</phase>
+                     </execution>
+                  </executions>
+                  <configuration>
+                     <buildArgs>
+                        -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager
+                        --enable-url-protocols=http
```
[the remark this answers is not in the document]
Comment (ryanemerson) [reply graalvm/client-hotrod/pom.xml:189]: Nope, removed.
